os/mm/mm_heap/mm_heapinfo_group.c

```diff
@@ -0,0 +1,189 @@
+				/* calculate the summation of stacks */
+				heapinfo_group[group_num].stack_size = 0;
+				for (stack_pid = 0; stack_pid < HEAPINFO_THREAD_NUM; stack_pid++) {
+					if (group_info[stack_pid].pid != -1 && group_info[stack_pid].group == group_num) {
+						heapinfo_group[group_num].stack_size += group_info[stack_pid].stack_size;
+					}
+				}
+				heapinfo_group[group_num].heap_size = heapinfo_group[group_num].peak_size - heapinfo_group[group_num].stack_size;
+			}
+			break;
+		}
+	}
+}
+
+/****************************************************************************
+ * Name: heapinfo_update_group_info
+ *
+ * Description:
+ * when create or release task/thread, check that the task/thread is 
+ * in group list
+ ****************************************************************************/
+void heapinfo_update_group_info(pid_t pid, int group, int type)
+{
+	int info_idx;
+	struct tcb_s *tcb;
+	switch (type) {
+	case HEAPINFO_INIT_INFO:
+		for (info_idx = 0; info_idx < HEAPINFO_THREAD_NUM; info_idx++) {
+			group_info[info_idx].pid = pid;
+			group_info[info_idx].group = group;
+			group_info[info_idx].stack_size = 0;
+		}
+		break;
+	case HEAPINFO_ADD_INFO:
+		if (group > max_group) {
+			max_group = group;
+		}
+		for (info_idx = 0; info_idx < HEAPINFO_THREAD_NUM; info_idx++) {
+			if (group_info[info_idx].pid <= 0) {
+				group_info[info_idx].pid = pid;
+				group_info[info_idx].group = group;
+				tcb = sched_gettcb(pid);
+				if (tcb) {
+					group_info[info_idx].stack_size = tcb->adj_stack_size;
+					heapinfo_update_group(tcb->adj_stack_size, pid);
+				}
+				break;
+			}
+		}
+		break;
+	case HEAPINFO_DEL_INFO:
+		for (info_idx = 0; info_idx < HEAPINFO_THREAD_NUM; info_idx++) {
+			if (pid == group_info[info_idx].pid) {
+				heapinfo_update_group((-1) * group_info[info_idx].stack_size, pid);
+
+				group_info[info_idx].pid = -1;
+				group_info[info_idx].group = -1;
+				group_info[info_idx].stack_size = 0;
+				break;
+			}
+		}
+		break;
+	default:
+		break;
+	}
+}
+
+/****************************************************************************
+ * Name: heapinfo_check_group_list
+ *
+ * Description:
+ * check that task/thread is in group list
+ ****************************************************************************/
+void heapinfo_check_group_list(pid_t pid, char *name)
+{
+	char *thread_list = CONFIG_HEAPINFO_USER_GROUP_LIST;
+
+	int group_num = 0;
+
+	char *name_start;
+	char *name_end;
+	int name_length;
+	name_start = name_end = thread_list;
+
+	while (*name_start != '\0') {
+		if (*name_end == '/' || *name_end == ',' || *name_end == '\0') {
+			name_length = name_end - name_start;
+			if (name_length == strlen(name)) {
+				if (strncmp(name_start, name, name_end - name_start) == 0) {
+					heapinfo_update_group_info(pid, group_num, HEAPINFO_ADD_INFO);
+					break;
+				}
+			}
+			if (*name_end == '/') {
+				group_num++;
+			} else if (*name_end == '\0') {
+				name_start = name_end;
+				continue;
```

Comment:
This will break loop?

Reply:
Loop will be terminated by while condition.

Reply:
my mistake, I see.